File: solutions/ocaml/068-task-scheduler.ml
(*
  Problem 68: Task Scheduler (LeetCode 621)
  Difficulty: Med
  Language: OCaml
*)
let least_interval tasks n =
  let freq = Hashtbl.create 26 in
  List.iter (fun c ->
    let v = try Hashtbl.find freq c with Not_found -> 0 in
    Hashtbl.replace freq c (v + 1)
  ) tasks;
  let freqs = Hashtbl.fold (fun _ v acc -> v :: acc) freq [] in
  let max_freq = List.fold_left max 0 freqs in
  let max_count = List.length (List.filter ((=) max_freq) freqs) in
  max (List.length tasks) ((max_freq - 1) * (n + 1) + max_count)

File: solutions/rust/068-task-scheduler.rs
/**
 * Problem 68: Task Scheduler (LeetCode 621)
 * Difficulty: Med
 * Language: Rust
 */
fn least_interval(tasks: &[char], n: i32) -> i32 {
    let mut freq = [0i32; 26];
    for &t in tasks { freq[(t as u8 - b'A') as usize] += 1; }
    let max_freq = *freq.iter().max().unwrap();
    let max_count = freq.iter().filter(|&&f| f == max_freq).count() as i32;
    (tasks.len() as i32).max((max_freq - 1) * (n + 1) + max_count)
}
